packages/persoo/src/Model/Product/ProductSearchResultsProvider.php

```diff
@@ -77,21 +77,47 @@ public function getProductsSearchResults(
         $after = $argument['after'] ?? null;
 
         $connectionBuilder = new ConnectionBuilder();
-        $offset = $connectionBuilder->getOffsetWithDefault($after, 0);
+        $page = $this->getPageFromOffsetAndLimit($connectionBuilder->getOffsetWithDefault($after, 0), $limit);
         $persooFilter = $this->productFilterToPersooFilterMapper->mapForSearch($productFilterData, $this->domain);
-
-        $result = $this->client->getData($search, PersooClient::PERSOO_INDEX_PRODUCTS, PersooClient::PERSOO_ACTION_SEARCH, $offset, $limit, $persooFilter);
+        $result = $this->client->getData($search, PersooClient::PERSOO_INDEX_PRODUCTS, PersooClient::PERSOO_ACTION_SEARCH, $page, $limit, $persooFilter);
 
         return $this->productConnectionFactory->createConnectionForAll(
             function ($offset, $limit) use ($result) {
                 $filterQuery = $this->filterQueryFactory->createSellableProductsByProductIdsFilter($result->getIds(), $limit);
+                $sortedProducts = $this->productElasticsearchRepository->getSortedProductsResultByFilterQuery($filterQuery)->getHits();
 
-                return $this->productElasticsearchRepository->getSortedProductsResultByFilterQuery($filterQuery)->getHits();
+                return $this->appendEmptyArrayValuesForCorrectFunctionOfConnectionBuilder($sortedProducts, $offset);
             },
             $result->getItemsCount(),
             $argument,
             $productFilterData,
             $this->productOrderingModeProvider->getOrderingModeFromArgument($argument),
         );
     }
+
+    /**
+     * @param int $getOffsetWithDefault
+     * @param int $limit
+     * @return int
+     */
+    protected function getPageFromOffsetAndLimit(int $getOffsetWithDefault, int $limit): int
+    {
+        return (int)ceil($getOffsetWithDefault / $limit);
+    }
+
+    /**
+     * @param array $sortedProducts
+     * @param int $offset
+     * @return array
+     */
+    protected function appendEmptyArrayValuesForCorrectFunctionOfConnectionBuilder(
```

Comment:
this is really disturbing :thinking: I do not understand it, could we please discuss it?